Themis/ViewModels/CorrectionSidebar/ProblemStatementCellViewModel.swift

```diff
@@ -32,47 +32,91 @@ class ProblemStatementCellViewModel: ObservableObject {
     }
 
-    func deleteTestCases(_ problemStatement: String) -> String {
-        return problemStatement
-            .replacingOccurrences(of: "[task][", with: "") // to remove symbol
-            .replacingOccurrences(of: "](.*())", with: "", options: .regularExpression) // to remove tests
+    private func replaceProblemStatementTestCases(_ problemStatement: String, _ feedbacks: [AssessmentFeedback]) -> String {
+        var problemStatementText = problemStatement
+
+        for feedback in feedbacks {
+            if let feedbackText = feedback.text {
+                problemStatementText = problemStatementText
+                    .replacingOccurrences(of: "\(feedbackText)", with: feedback.credits == 0.0 ? "0" : "1")
+            }
+        }
+
+        let lines = problemStatementText.components(separatedBy: "\n")
+        problemStatementText = ""
+
+        for var line in lines {
+            if line.contains("[task][") && line.contains("0") {
+                line = calculateScoreOfPassingTests(line)
+                line = line.replacingOccurrences(of: "[task][", with: " ![Test failed](asset:///TestFailedSymbol) **")
+            } else if line.contains("[task][") && line.contains("1") {
+                line = calculateScoreOfPassingTests(line)
+                line = line.replacingOccurrences(of: "[task][", with: " ![Test passed](asset:///TestPassedSymbol) **")
+            }
+
+            problemStatementText.append(line + "\n")
+        }
+
+        return problemStatementText
     }
 
-    func replaceTestsColor(_ problemStatement: String, _ colorSchemeVariable: ColorScheme) -> String {
-        return problemStatement
-                .replacingOccurrences(of: "testsColor\\([A-z]+\\)", with: colorSchemeVariable == .light ? "black" : "white", options: .regularExpression)
+    private func calculateScoreOfPassingTests(_ line: String) -> String {
+        if let rangeStart = line.range(of: "](", range: line.startIndex..<line.endIndex) {
+            let lineBeforeScore = String(line[...rangeStart.lowerBound].dropLast())
+
+            let scoreSubstring = String(line[rangeStart.lowerBound...].dropFirst().dropFirst().dropLast())
+
+            let scoreSubstringArray = scoreSubstring.components(separatedBy: ",").map { Int($0)! }
+
+            let lineScore = " (\(scoreSubstringArray.reduce(0, +)) of \(scoreSubstringArray.count) tests passing)** "
+
+            return lineBeforeScore + lineScore
+        }
+
+        return line
+    }
+
+    private func replaceUMLTestsColor(_ problemStatement: String, _ feedbacks: [AssessmentFeedback], _ colorSchemeVariable: ColorScheme) -> String {
+        var plantUML = problemStatement
+
+        for feedback in feedbacks {
+            if let feedbackText = feedback.text {
+                plantUML = plantUML.replacingOccurrences(
+                    of: "testsColor(\(feedbackText))",
+                    with: feedback.credits == 0.0 ? "red" : "green")
+            }
+        }
```

Comment:
Also could be in a ViewModel?

Reply:
It is in a ViewModel (ProblemStatementCellViewModel).